draft-specifications/Notation.md

```diff
@@ -0,0 +1,218 @@
+- first - last
+- optional ?
+- ( group )
+- _reference_
+- _referent_ : _definition_
+- :: semantics
+
+Both grammars ignore arbitrary [ASCII](https://datatracker.ietf.org/doc/html/rfc20)
+space (SP), tab (HT), carriage return (CR), and line feed (LF) characters between
+tokens.
+
+Commentary in block quotes is not normative.
+
+> Non-normative commentary.
+
+## Value
+
+A representation of any expressible value in an OCapN message.
+
+- _abstract-value_: _abstract-boolean_ / _abstract-integer_ /
+  _abstract-float64_ / _abstract-selector_ / _abstract-string_ /
+  _abstract-byte-array_ / _abstract-struct_ /
+  _abstract-list_ / _abstract-record_
+- _concrete-value_: _concrete-boolean_ / _concrete-integer_ /
+  _concrete-float64_ / _concrete-selector_ / _concrete-string_ /
+  _concrete-byte-array_ / _concrete-struct_ /
+  _concrete-list_ / _concrete-record_
+
+## Boolean
+
+A value that is either true or false.
+
+> Examples:
+> - `f` corresponds to `f`.
+> - `t` corresponds to `t`.
+
+- _abstract-boolean_: `f` / `t` :: Corresponding to false and true respectively.
+- _concrete-boolean_: `f` / `t`
+
+## Integer
+
+An arbitrary precision signed integer.
+
+> Examples:
+> - `42` corresponds to `42+`.
+> - `-1` corresponds to `1-`.
+> - `0` corresponds to `0+`.
+
+- _abstract-integer_: _sign_? _integer-digits_
+- _sign_: `+` / `-`
+- _integer-digits_: ( `0` / ( `1` - `9` ) _digit_* )
+- _concrete-integer_: _integer-digits_ _sign_
+
+## Float64
+
+An IEEE 754 64-bit floating point number.
+
+> Examples:
+> - `nan` corresponds to the bytes 44 (`"D"`), 7f, f8, 00, 00, 00, 00, 00, 00
+>   in hexadecimal.
+
+- _abstract-float64_: _abstract-float64-number_ / ( _sign_? `inf` ) / `nan`
+- _abstract-float64-number_: _abstract-integer_ `.` _digit_* /
+  _sign_? `.` _digit_+ :: Corresponding to the nearest expressible concrete IEEE
+  754 64-bit floating point number, rounding ties to even.
+- _concrete-float64_: `D` followed by the corresponding 8 bytes of an IEEE 754
+  64-bit floating point number.
+
+## String
+
+A sequence of Unicode code points excluding surrogates (U+D800-U+DFFF).
+
+> Example: `"twine"` corresponds to `5"twine`.
+
+- _abstract-string_: `"` _abstract-character_ * `"`
+- _abstract-character_:: _any printable ASCII character except `"` or `\`_ ::
+  note that spaces are printable
+- _concrete-string_: _length_ `"` _bytes_
+- _length_: _integer-digits_ :: The number of bytes in _bytes_ as ASCII decimal
+  digits.
+- _bytes_: _byte_* :: The bytes of the string in UTF-8 encoding.
+
+We do not attempt to capture strings with embedded quotes or non-ASCII Unicode
+characters in the abstract notation, but all Unicode strings excluding
+surrogates (U+D800-U+DFFF) are expressible in the concrete representation.
+
+## Selector
+
+A sequence of Unicode code points excluding surrogates (U+D800-U+DFFF).
+
+> Example: `'fleur-de-lis` corresponds to `12'fleur-de-lis`.
+
+- _abstract-selector_: `'` _name_
+- _name_: _alpha_ ( _alpha_ / _digit_ / `-` )*
```

Comment:
I suspect this would suffice for our uses in the other documents, but I'm wondering why we're limiting the name to starting with an alpha character, limited to the entire selector to contain just alphanumeric and `-`? I'd have expected a definition more similar to strings.

Reply:
I am adding colon (`:`) since it’s an obvious omission. The concrete syntax allows the full range of strings, but I figure it’s not worth the trouble to define that exhaustively for the abstract notation given our usage in examples doesn’t transcend this subset. I specifically wish to avoid (for now) the escaping rules necessary to express the entire range. That is, all kinds of space and single quote, at least. So, given the choice between extremes: covering all expressible symbols versus covering only expressed symbols, I took the lazy option.